src/saveload/saveload.cpp

```diff
@@ -1228,125 +1253,103 @@ class SlStorageHelper {
-	*/
-	static void SlSaveLoad(void *storage, VarType conv)
+	 * @param cmd The SaveLoadType ware are saving/loading.
+	 */
+	static void SlSaveLoad(void *storage, VarType conv, SaveLoadType cmd = SL_VAR)
 	{
+		assert(cmd == SL_VAR || cmd == SL_REF);
+
 		SlStorageT *list = (SlStorageT *)storage;
 
 		switch (_sl.action) {
-			case SLA_SAVE: {
+			case SLA_SAVE:
 				SlWriteUint32((uint32)list->size());
 
 				for (auto &item : *list) {
-					SlSaveLoadConv(&item, conv);
+					SlSaveLoadMember(cmd, &item, conv);
 				}
 				break;
-			}
+
 			case SLA_LOAD_CHECK:
 			case SLA_LOAD: {
 				size_t length = SlReadUint32();
 
 				/* Load each value and push to the end of the storage. */
 				for (size_t i = 0; i < length; i++) {
 					Tvar data;
-					SlSaveLoadConv(&data, conv);
+					SlSaveLoadMember(cmd, &data, conv);
 					list->push_back(data);
 				}
 				break;
 			}
+
 			case SLA_PTRS:
+				for (auto &item : *list) {
+					SlSaveLoadMember(cmd, &item, conv);
+				}
 				break;
+
 			case SLA_NULL:
 				list->clear();
 				break;
+
 			default: NOT_REACHED();
 		}
 	}
 };
 
 /**
- * Return the size in bytes of a list
- * @param list The std::list to find the size of
+ * Return the size in bytes of a list.
+ * @param list The std::list to find the size of.
+ * @param conv VarType type of variable that is used for calculating the size.
  */
-static inline size_t SlCalcListLen(const void *list)
+static inline size_t SlCalcListLen(const void *list, VarType conv)
 {
-	const std::list<void *> *l = (const std::list<void *> *) list;
-
-	int type_size = IsSavegameVersionBefore(SLV_69) ? 2 : 4;
-	/* Each entry is saved as type_size bytes, plus type_size bytes are used for the length
-	 * of the list */
-	return l->size() * type_size + type_size;
+	return SlStorageHelper<std::list, void *>::SlCalcLen(list, conv, SL_REF);
 }
 
 /**
  * Save/Load a list.
- * @param list The list being manipulated
- * @param conv SLRefType type of the list (Vehicle *, Station *, etc)
+ * @param list The list being manipulated.
+ * @param conv VarType type of variable that is used for calculating the size.
  */
-static void SlList(void *list, SLRefType conv)
+static void SlList(void *list, VarType conv)
 {
 	/* Automatically calculate the length? */
 	if (_sl.need_length != NL_NONE) {
-		SlSetLength(SlCalcListLen(list));
+		SlSetLength(SlCalcListLen(list, conv));
 		/* Determine length only? */
 		if (_sl.need_length == NL_CALCLENGTH) return;
 	}
 
-	typedef std::list<void *> PtrList;
-	PtrList *l = (PtrList *)list;
-
-	switch (_sl.action) {
-		case SLA_SAVE: {
-			SlWriteUint32((uint32)l->size());
-
-			PtrList::iterator iter;
-			for (iter = l->begin(); iter != l->end(); ++iter) {
-				void *ptr = *iter;
-				SlWriteUint32((uint32)ReferenceToInt(ptr, conv));
-			}
-			break;
-		}
-		case SLA_LOAD_CHECK:
-		case SLA_LOAD: {
-			size_t length = IsSavegameVersionBefore(SLV_69) ? SlReadUint16() : SlReadUint32();
```

Comment:
I think save games before SLV_69 are broken by this. The new code always reads 32 bits.
The same then also holds for the SlCalc(List)Len.

Since SLE_CONDLST was added in SLV_57, opntitle.dat just works fine. PSG-40 is "broken" (unexpected end of chunk) in this PR, with 1.11.0 it's not broken.

Reply:
wauw, that took way to long to decypher what you said here ;)

Seems in all places the length-thing did survive except for SLA_LOAD. Ironical :)

`SlCalc` etc does have the right code. But, ironically, `SlCalc` doesn't need it, as that is only used for writing. But not going to change everything all at once ;)

Reply:
Should be fixed; can you try the savegame that was failing?

I have a wide collection, but clearly not one that triggered this issue :D